File: Gen2_0_PP/Assignment/leetcode_permute.cpp
class Solution {
public:
    
    static void rev(vector<int> &nums,int i,int j){
        while(i < j){
            swap(nums[i],nums[j]);
            i++;j--;
        }
    }
    
    static void LR(vector<int> &nums,int i,int j){
        rev(nums,i,j-1);
        rev(nums,i,j);
    }
    
    static void RR(vector<int> &nums,int i,int j){
        rev(nums,i+1,j);
        rev(nums,i,j);
    }
    
    static void func(vector<int>& nums,int i,vector<vector<int> > &ans){
        if(i == nums.size()-1){
            ans.push_back(nums);
        }
        
        for(int j = i; j < nums.size(); j++){
            LR(nums,i,j);
            func(nums,i+1,ans);
            RR(nums,i,j);
            
        }
    }
    
    vector<vector<int>> permute(vector<int>& nums) {
        
        vector<vector<int> > ans;
        //vector<int> temp;
        func(nums,0,ans);
        return ans;
        
    }
};

File: Gen2_0_PP/Assignment/leetcode_permute.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        ans=[]
        def perm(nums,start):
            if start==len(nums)-1:
                ans.append(nums) 
                return
            for j in range(start,len(nums)):
                nums[j],nums[start]=nums[start],nums[j] 
                perm(nums[:],start+1)  
        perm(nums,0)
        return ans

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        if n < 2:
            yield nums
        elif n == 2:
            yield from [nums, nums[::-1]]
        else:
            for i, k in enumerate(nums):
                rem = nums[:i] + nums[i+1:]
                for perm in self.permute(rem):
                    yield [k] + perm